File: Day18-BinaryTree5/112. Path Sum/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> bool:
        if not root:
            return False
        if not root.left and not root.right:
            return targetSum == root.val
        return self.pathSum(root.left, targetSum - root.val) or self.pathSum(root.right, targetSum - root.val)
    
# Test
root = TreeNode(5, TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2))), TreeNode(8, TreeNode(13), TreeNode(4, TreeNode(5), TreeNode(1))))
targetSum = 22
sol = Solution()
print(sol.pathSum(root, targetSum)) # True

File: Day18-BinaryTree5/112. Path Sum/main.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(NULL), right(NULL) {}
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


TreeNode* constructTree(const vector<int>& values, int index) {
    if (index >= values.size() || values[index] == -1) {
        return nullptr;
    }
        
    TreeNode* root = new TreeNode(values[index]);
    root->left = constructTree(values, 2 * index + 1);
    root->right = constructTree(values, 2 * index + 2);
        
    return root;
}

class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(root == NULL) return false;
        return traversal(root, targetSum);
    }

    bool traversal(TreeNode* root, int targetSum){
        if(root->left == NULL && root->right == NULL && targetSum == 0) {
            return true;
        }
        if(root->left == NULL && root->right == NULL) {
            return false;
        }
        if(root->left){
            if(hasPathSum(root->left, targetSum - root->left->val)){
                return true;
            }
        }
        if(root->right){
            if(hasPathSum(root->right, targetSum - root->right->val)){
                return true;
            }
        }
        return false;
    }
};

int main(){
    // Usage:
    vector<int> treeValues = {5, 4, 8, 11, 7, 2, 13, -1, -1, -1, -1, -1, -1, 4, 1};
    TreeNode* root = constructTree(treeValues, 0);
    Solution s;
    cout << s.hasPathSum(root, 22);
    return 0;
}